src/main.cpp

```diff
@@ -7626,70 +7626,74 @@ bool SendMessages(CNode *pto)
-                        GetTime() - pto->nTimeConnected);
-                    continue;
-                }
+                int invsz = pto->vInventoryToSend.size();
+                vInvSend.reserve(invsz);
+                // about 3/4 of the nodes should end up in this list, so over-allocate by 1/10th + 10 items
+                vInvWait.reserve((invsz * 3) / 4 + invsz / 10 + 10);
 
-                // trickle out tx inv to protect privacy
-                if (inv.type == MSG_TX && !fSendTrickle)
+                // Make copy of vInventoryToSend while cs_inventory is locked but also ignore some tx and defer others
+                for (const CInv &inv : pto->vInventoryToSend)
                 {
-                    // 1/4 of tx invs blast to all immediately
-                    static uint256 hashSalt;
-                    if (hashSalt.IsNull())
-                        hashSalt = GetRandHash();
-                    uint256 hashRand = ArithToUint256(UintToArith256(inv.hash) ^ UintToArith256(hashSalt));
-                    hashRand = Hash(BEGIN(hashRand), END(hashRand));
-                    bool fTrickleWait = ((UintToArith256(hashRand) & 3) != 0);
-
-                    if (fTrickleWait)
+                    if (inv.type == MSG_TX)
                     {
-                        vInvWait.push_back(inv);
-                        continue;
+                        if (chokeTxInv)
+                            continue;
+                        // skip if we already know abt this one
+                        if (pto->filterInventoryKnown.contains(inv.hash))
+                            continue;
+                        if (fSendTrickle)
+                        {
+                            // 1/4 of tx invs blast to all immediately
+                            if ((insecure_rand() & 3) != 0)
+                            {
+                                vInvWait.push_back(inv);
+                                continue;
+                            }
+                        }
                     }
+                    vInvSend.push_back(inv);
+                    pto->filterInventoryKnown.insert(inv.hash);
                 }
+                pto->vInventoryToSend = vInvWait;
+            }
 
-                pto->filterInventoryKnown.insert(inv.hash);
-
-                vInv.push_back(inv);
-                if (vInv.size() >= 1000)
+            const int MAX_INV_ELEMENTS = 1000;
+            int sz = vInvSend.size();
+            if (sz)
+            {
+                LOCK(pto->cs_vSend);
+                for (int i = 0; i < sz; i += MAX_INV_ELEMENTS)
                 {
-                    pto->PushMessage(NetMsgType::INV, vInv);
-                    vInv.clear();
+                    int sendsz = std::min(MAX_INV_ELEMENTS, sz - i);
+                    std::vector<CInv> vInv(sendsz);
+                    for (int j = 0; j < sendsz; j++)
+                        vInv[j] = vInvSend[i + j];
```

Comment:
aren't we introducing an O(n^2) loop here?

Reply:
no because i is incremented by MAX_INV_ELEMENTS, and sendsz is always <= MAX_INV_ELEMENTS.

This code is just sending the vInvSend vector out in chunks of size MAX_INV_ELEMENTS or smaller.  We need to copy what we are sending into a new vector because you can't specify a subset of the vector into PushMessage.  As the TODO below says, if we add that feature, we can remove the above for loop.

Reply:
thanks for the explanation. makes sense.